### PR TITLE
Add writable function calls for issue create, update, and comment!

### LINKED ISSUE #10: implement writable function calls
lets continue to test the limits of function calling by adding some basic writable functions such as createIssue, createFile, createCommit, etc

### PR DESCRIPTION
**Pull request by skylar-anderson**
closes #10 by adding writable functions for issue create, edit, and comment.

### CODE REVIEW

**Comment by skylar-anderson** [comment]
### Code Review for PR #21

After reviewing the changes, here are some concerns and suggestions:

- **Function Implementations (createIssue, createIssueComment, updateIssue):** It's great that asynchronous patterns with `await` are utilized for the function implementations. However, I noticed that the error handling could potentially be improved. Instead of directly returning the result of `await`ing the function calls, consider wrapping them in a `try/catch` block. This would allow for more graceful error handling and provide clearer error messages if the GitHub API calls fail.

- **Consistency in Parameters and Documentation:** All new function implementations have correctly documented parameters, which is good for maintainability. Yet, some methods seem to have inconsistent parameter names compared to their usage. For instance, `createIssueComment.run` has a parameter named `issueNumber`, but when you call `createIssueComment`, you use `args["issueNumber"]`. This could be problematic if the structure of `args` changes. You might want to destructure parameters early on for more robust code.

- **Routing Logic in `listIssues`:** I appreciate the effort to support filtering issues by label in the `listIssues` function, but it seems you forgot to actually push the label filter into the `filters` array (line `listIssues.ts:60`). This oversight makes the label filtering non-functional.

- **Unused `is` parameter in `listIssues.run`:** It seems `is: "issue" | "pull-request" = "issue"` parameter isn't actually used within the `listIssues.run` function. It's defined twice in the function signature which is a code smell and likely a merging mistake.

- **TypeScript Types Usage:** Although you've added TypeScript type definitions, which is excellent for code robustness, it seems you're not utilizing TypeScript to its full power. Specifically, consider using discriminated unions or similar patterns for complex conditional logic based on parameters—this helps prevent runtime errors.

In summary, while the functionality introduced in this PR is quite valuable for making issue management more automated, there are areas where the code could be more robust, maintainable, and consistent. Integrating the suggested changes can significantly improve the quality of the codebase.

**Comment by skylar-anderson** [comment]
After re-reviewing the latest code changes in PR #21, here are some focused critiques to enhance the code quality:

1. **Error Handling:** The new function implementations such as `createIssue`, `createIssueComment`, and `updateIssue` could benefit from structured error handling. It's advisable to include `try/catch` blocks around `await` calls to manage exceptions effectively and provide meaningful error messages to the calling context.

2. **Parameter Consistency:** Ensure that the parameters are consistent across the codebase. The `createIssueComment.run` uses `args["issueNumber"]` which could lead to errors if the structure of `args` changes. Consider deconstructing the parameters or validating that the keys match their intended use cases.

3. **Filtering Logic:** The fix for filtering labels in `listIssues` is necessary since the labels are not being pushed into the filters array. This change will enable label-based issue filtering as intended.

4. **Unused Parameters:** Review the code to remove any unused parameters, such as the redundant `is` parameter in `listIssues.run`. Keeping parameters that are not used contributes to confusion and potential inconsistencies.

5. **Code Simplicity and Readability:** Aim for simplicity and clarity in your code. There is room to streamline some of the logic and remove redundant or confusing parts of the implementation. This includes cleaning up enums and aligning function signatures with their purpose.

6. **Utilizing TypeScript Effectively:** Enhance the robustness of your code by making full use of TypeScript features like discriminated unions. This adds compile-time checks that can prevent common errors related to conditional logic.

Overall, focusing on these aspects can lead to improved code robustness, consistency, and maintainability. It's vital to iterate and refine where possible before finalizing the PR.

**Comment by skylar-anderson** [comment]
I've identified an area for improvement regarding the error handling mechanism in your `createIssue` endpoint. Here is an improved version with more robust practices:

```typescript
// app/api/issues/route.ts

export async function POST(request: NextRequest) {
  ...
  try {
    const issue = await createIssue({
      repository,
      title,
      body,
      labels,
      assignees,
    });
    return NextResponse.json({ status: "success", issue });
  } catch (error) {
    console.error(`Failed to create issue: ${error.message}`);
    return new Response(JSON.stringify({
      status: "error",
      message: "An error occurred while attempting to create the issue.",
      details: error instanceof Error ? error.message : String(error),
    }), {
      status: 500,
      headers: { 'Content-Type': 'application/json' },
    });
  }
}
```

Changes and justifications:
- Replaced `console.log` with `console.error` for more semantic significance of logging errors.
- Used `error.message` for a more specific error detail logging.
- Respond with a `Response` object instead of a plain text string to set the proper response code and headers.
- Include a structured error response with `message` and `details` to provide clearer insights to the client when errors occur.
- Check `instanceof Error` for more reliable error type checking.

By implementing these error handling improvements, your code will not only provide better logging for server-side debugging but will also enhance the client-side handling of errors, making the API more reliable and easier to work with for the users.
